lib/Dialect/Torch/Transforms/GlobalizeObjectGraph.cpp

```diff
@@ -142,11 +143,36 @@ class ObjectGraphInfo {
       if (slotIt == nameToSlot.end())
         op->emitError() << "Reference to non-existing module slot " << slotName
                         << "in " << moduleType.getClassName();
-      usedSlots.insert(slotIt->getValue());
+      slotUses[slotIt->getValue()].insert(op);
     });
     return success();
   }
 
+  LogicalResult checkUsesForValueSemantics(Value root) {
+    for (auto &use : root.getUses()) {
+      // For non-slot operations, simply trace the value's uses.
+      if (!isa<SlotOp>(use.getOwner()))
+        return checkUsesForValueSemantics(use.get());
+
+      // For `torch.slot` operations, lookup uses of the slot using the
+      // `slotUses` map, before checking whether each use is either free of side
+      // effects (e.g. a list constructor) or has value semantics.
+      auto traitCheck = [](Operation *operation) {
+        return llvm::all_of(operation->getUses(), [](OpOperand &operand) {
+          auto owner = operand.getOwner();
+          return mlir::isSideEffectFree(owner) ||
```

Comment:
isSideEffectFree isn't the property we want -- an op that isSideEffectFree can still create aliases. Can you check only for HasValueSemantics here?

Reply:
I see.  The problem is that I have a `torch.prim.TupleConstruct` (see updated test in `//test/Dialect/Torch/GlobalizeObjectGraph/initializers.mlir`) but since that op doesn't have the `HasValueSemantics` trait associated with it, the guard condition fails.  Are list / tuple constructors safe users of `GetAttr`s?

Reply:
No, they are not safe. You could read out the tensor from the Tuple or List and then mutate it.

Reply:
Oh, that's right.  Makes sense now, thanks!